community/primitive-collections/src/main/java/org/neo4j/concurrent/AsyncEvents.java

```diff
@@ -0,0 +1,162 @@
+    private volatile AsyncEvent stack; // Accessed via AtomicReferenceFieldUpdater
+    private volatile boolean shutdown;
+
+    /**
+     * Construct a new {@code AsyncEvents} instance, that will use the given consumer to process the events.
+     *
+     * @param eventConsumer The {@link Consumer} used for processing the events that are sent in.
+     */
+    public AsyncEvents( Consumer<T> eventConsumer )
+    {
+        endSentinel = new EndEvent();
+        shutdownSentinel = new ShutdownEvent();
+        this.eventConsumer = eventConsumer;
+        stack = endSentinel;
+    }
+
+    @Override
+    public void send( T event )
+    {
+        AsyncEvent prev = updater.getAndSet( this, event );
+        assert prev != null;
+        event.next = prev;
+        if ( prev == endSentinel )
+        {
+            LockSupport.unpark( endSentinel.runner );
+        }
+        else if ( prev == shutdownSentinel )
+        {
+            AsyncEvent events = updater.getAndSet( this, shutdownSentinel );
+            process( events );
+        }
+    }
+
+    @Override
+    public void run()
+    {
+        endSentinel.runner = Thread.currentThread();
+
+        do
+        {
+            AsyncEvent events = updater.getAndSet( this, endSentinel );
+            process( events );
+
+            if ( stack == endSentinel && !shutdown )
+            {
+                LockSupport.park( this );
+            }
+        }
+        while ( !shutdown );
+
+        AsyncEvent events = updater.getAndSet( this, shutdownSentinel );
+        process( events );
+    }
+
+    private void process( AsyncEvent events )
+    {
+        events = reverseAndStripEndMark( events );
+
+        while ( events != null )
+        {
+            @SuppressWarnings( "unchecked" )
+            T event = (T) events;
+            eventConsumer.accept( event );
+            events = events.next;
+        }
+
+        endSentinel.next = null;
+        shutdownSentinel.next = null;
+    }
+
+    private AsyncEvent reverseAndStripEndMark( AsyncEvent events )
+    {
+        AsyncEvent result = null;
+        while ( events != endSentinel && events != shutdownSentinel )
+        {
+            AsyncEvent tmp;
+            do
+            {
+                tmp = events.next;
+            }
+            while ( tmp == null );
+            events.next = result;
+            result = events;
+            events = tmp;
+        }
+        return result;
+    }
+
+    /**
+     * Initiate the shut down process of this {@code AsyncEvents} instance.
+     *
+     * This call does not block or otherwise wait for the background thread to terminate.
+     */
+    public void shutdown()
+    {
+        shutdown = true;
+        LockSupport.unpark( endSentinel.runner );
+    }
+
+    private static final class EndEvent extends AsyncEvent
+    {
+        volatile Thread runner;
+    }
+
+    private static final class ShutdownEvent extends AsyncEvent
+    {
+    }
```

Comment:
It's a technique already in use in `WorkSync` so it's not new. WorkSync is used in 2.2 and 2.3 during the commit process. We could unify them though.